### ISSUE TITLE
Deprecation: #104778 - Instantiation of IconRegistry in ext_localconf.php

### ISSUE THREAD

**Issue by simonschaufi** [sabbelasichon/typo3-rector#4344]
Deprecation: #104778 - Instantiation of IconRegistry in ext_localconf.php

https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/13.3/Deprecation-104778-InstantiationOfIconRegistryInExtLocalconf.html

# Deprecation: \#104778 - Instantiation of IconRegistry in ext_localconf.php

See [104778](https://forge.typo3.org/issues/104778)

## Description

Since TYPO3 v11 it is possible to automatically register own icons via
`Configuration/Icons.php`. Prior to this, extension authors used to
register icons manually via instantiating the
php:<span class="title-ref">TYPO3CMSCoreImagingIconRegistry</span> in
their `ext_localconf.php` files. This method has now been deprecated. It
is recommended to switch to the newer method introduced with [94692](https://forge.typo3.org/issues/94692).

## Impact

Instantiating `\TYPO3\CMS\Core\Imaging\IconRegistry` inside
`ext_localconf.php` files will trigger a deprecation-level log entry.

## Affected installations

All installations, which instantiate
`\TYPO3\CMS\Core\Imaging\IconRegistry` before the
`\TYPO3\CMS\Core\Core\Event\BootCompletedEvent`. This includes
`ext_localconf.php` files as well as `TCA/Overrides`.

## Migration

The most common use-cases can be accomplished via the
`Configuration/Icons.php` file.

Before:

``` php
<?php

$iconRegistry = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance(
  \TYPO3\CMS\Core\Imaging\IconRegistry::class,
);
$iconRegistry->registerIcon(
    'example',
    \TYPO3\CMS\Core\Imaging\IconProvider\SvgIconProvider::class,
    [
        'source' => 'EXT:example/Resources/Public/Icons/example.svg'
    ],
);
```

After:

``` php
<?php

return [
    'example' => [
        'provider' => \TYPO3\CMS\Core\Imaging\IconProvider\SvgIconProvider::class,
        'source' => 'EXT:example/Resources/Public/Icons/example.svg',
    ],
];
```

For more complex tasks, it is recommended to register an event listener
for the `\TYPO3\CMS\Core\Core\Event\BootCompletedEvent`. At this stage
the system is fully booted and you have a completely configured
IconRegistry at hand.

In case the registry was used in `TCA/Overrides` files to retrieve icon
identifiers, then this should be replaced completely with static
identifiers. The reason behind this is, that the registry isn't even
fully usable at this stage. TCA isn't fully built yet and icons can
still be registered at a later point.

<div class="index">

PHP-API, NotScanned, ext:core

</div>

**Comment by christophlehmann** [comment]
* [ ] When written to Icons.php the `iconProvider` value becomes `'TYPO3\\CMS\\Core\\Imaging\\IconProvider\\SvgIconProvider'` and not `\TYPO3\CMS\Core\Imaging\IconProvider\SvgIconProvider::class`

**Comment by simonschaufi** [comment]
Please create a new bug issue for this. Thank you.

**Comment by simonschaufi** [comment]
Implemented with #2659

**Comment by christophlehmann** [comment]
@simonschaufi The Rector in #2659 only replaces `typo3/sysext/core/Resources/Public/Icons/T3Icons/content/` with `typo3/sysext/core/Resources/Public/Icons/T3Icons/svgs/content/` so it's not completed

See: https://github.com/sabbelasichon/typo3-rector/pull/2659/files#diff-0696c545947019794bc6fa77e98e85153cd6bb201a4f7d956f73a50b11a6f8d9R68
